WIP - libcephfs: add async I/O capability

ffilz opened the pull request:
This is the first pass at supporting async I/O calls from Ganesha. I would appreciate another set of eyes to verify that I have appropriately covered all the bases. I am concerned that somehow I have left a path for synchronous completion which leaves confusion about the return value. The intent is to have all I/O complete async even if it completes before returning to the caller (which is prepared to deal with that situation). I'm also concerned about the best way to detect EOF. Ganesha currently assumes a 0 length read indicates EOF, but maybe we can do better.

Signed-off-by: Frank S. Filz <ffilzlnx@mindspring.com>

I haven't tried to complete the checklist, I know there is at least documentation that should be written, and tests are yet to be written.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [x] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

jtlayton commented:
AIUI, the only reliable way to detect the EOF in ceph is to keep track of the inode size and clamp your reads there. Because we're reading from an array of objects, a short read doesn't necessarily indicate an EOF condition. It may just be that the object has a sparse region at the end. Ditto for a non-existent OSD object.

jtlayton commented:
This patch is enormous. It'd be very nice to break this up into smaller changes, as it's a little hard for me to follow what your general approach is. That'll also be nice if this causes a regression and we have to bisect to find it.

jtlayton commented:
> That's odd. For an existing object, I'd expect the object size to be truncated to the one as defined in its file-layout to avoid a messy short read. Or, is that intentional ?

The OSD doesn't really have the concept of file layouts. Those are a cephfs concept and are managed by the MDS. All the OSD understands is objects, and multiple (usually 4M) objects back a single cephfs inode.

OSD objects can be sparse with byte granularity as well, so it really only stores what the clients have written. If the client leaves a sparse area in an object, or never creates an object that is completely sparse, then that's OK and is preferable since it takes up less space in the backing store. This is also what makes fscrypt possible -- we have to know which parts are actually sparse to properly decrypt the data.

The upshot is that cephfs clients must be prepared to zero-fill these holes in OSD data for the caller on a read. A hole could even be at the end of the file too which is why we have to carefully track the file size.

kotreshhr commented on src/client/Client.cc:
```diff
@@ -15587,6 +16187,9 @@ void Client::handle_conf_change(const ConfigProxy& conf,
     objectcacher->set_max_dirty_age(cct->_conf->client_oc_max_dirty_age);
   }
+  if (changed.count("client_block_writes_upfront")) {
+    objectcacher->set_block_writes_upfront(cct->_conf->client_oc_max_dirty_age);
```
The new config needs to added in the yaml mds-client.yaml.in

lxbsz commented on src/client/Client.cc:
```diff
@@ -10154,12 +10540,13 @@ int Client::pwritev(int fd, const struct iovec *iov, int iovcnt, int64_t offset)
                                        unsigned iovcnt, int64_t offset,
-                                       bool write, bool clamp_to_int)
+                                       bool write, bool clamp_to_int,
+                                       CC_Onfinish *onfinish)
```
You can just set `CC_Onfinish *onfinish=nullptr` and then there is no need to change the irrelevant code.

lxbsz commented on src/client/Client.cc:
```diff
@@ -9820,21 +9820,297 @@ int Client::preadv(int fd, const struct iovec *iov, int iovcnt, loff_t offset)
+  bool onuninlinefinished;
+  bool movepos;
+  bool try_complete();
+};
```
We'd better put this in the `Client.h`, the `Client.cc` is already very huge one ?

And also for `Client_Read_Sync_Async` ?

ajarr commented on src/client/Client.cc:
```diff
@@ -10017,36 +10396,43 @@ int Client::_read_async(Fh *f, uint64_t off, uint64_t len, bufferlist *bl)
+      crf->complete(r);
+      client_lock.lock();
+    } else {
+      get_cap_ref(in, CEPH_CAP_FILE_CACHE);
```
Why do this?

ajarr commented on src/client/Client.cc:
```diff
@@ -9823,18 +9824,314 @@ int Client::preadv(int fd, const struct iovec *iov, int iovcnt, loff_t offset)
+  bool iofinished;
+  void finish_io(int r);
+  void finish_onuninline(int r);
+  void finish_ui(int r);
```
what's does "finish_ui" do? I couldn't find this else where in the code.

ajarr commented on src/client/Client.cc:
```diff
@@ -10221,20 +10633,208 @@ int Client::_preadv_pwritev(int fd, const struct iovec *iov, unsigned iovcnt, in
+public:
+  void finish_io(int r);
+  void finish_onuninline(int r);
+  void finish_ui(int r);
```
what's does "finish_ui" do? I couldn't find this else where in the code.

ffilz commented:
I've updated this patch set. I hope I've addressed comments. I've also added a unit test and resolved several issues that arose from testing. I will shortly start testing from Ganesha.

adamemerson commented on src/client/Client.cc:
```diff
@@ -9738,6 +9738,7 @@ void Client::unlock_fh_pos(Fh *f)
 int Client::uninline_data(Inode *in, Context *onfinish)
 {
   if (!in->inline_data.length()) {
+    // TODO - should we drop the client_lock here?
```
@vshankar Can you weigh in here?

vshankar replied:
Sorry - I wasn't following this PR closely. I'll take a look next week.

mchangir replied:
although the `onfinish` context signals a condition variable on completion, its a no-op if there's no inline data to be uninlined, since the `onfinish` context is not handed over to an asynchronous network operation to signal is completion
so, IMO, we don't need to drop the `client_lock` here

ffilz replied:
Thanks. Removed the comment.

ffilz replied:
Also, onunline is now only done fro the write path AND is done synchronously, so all the async onunline bits that I had added in are gone.

ffilz commented on src/client/Client.cc:
```diff
@@ -9820,21 +9820,297 @@ int Client::preadv(int fd, const struct iovec *iov, int iovcnt, loff_t offset)
+  Context *onuninline;
+  Context *iofinish;
+  bool is_read_async;
+  int have;
```
Fair enough. I had just used what had been a local variable name that now has to be preserved in the object.

ffilz replied:
I think I've fixed this one

ffilz commented on src/client/Client.cc:
```diff
@@ -9820,21 +9820,297 @@ int Client::preadv(int fd, const struct iovec *iov, int iovcnt, loff_t offset)
+  int wanted;
+  bufferlist *bl;
+  Filer *filer;
+  int have;
```
Sure as above.

lxbsz reviewed (commented):
Could add the tags for each commit, something like:

`Client: XXX`
`test: xxx`
?

lxbsz commented on src/client/Client.h:
```diff
@@ -1222,6 +1229,207 @@ class Client : public Dispatcher, public md_config_obs_t {
+    }
+
+    void finish(int r) override {
+      //
```
remove this.

ffilz replied:
Sure

lxbsz commented on src/client/Client.h:
```diff
@@ -1222,6 +1229,207 @@ class Client : public Dispatcher, public md_config_obs_t {
+
+    void finish(int r) override {
+      CRF->finish_onuninline(r);
+      complete(r);
```
Won't this cause infinitely loop ? The `Context::complete()` will call the `finish()` again.

ffilz replied:
Oops, yes. That may have been a left over from the class deriving from C_SaferCond at one point during my development.

ffilz commented on src/test/client/async.cc:
```diff
@@ -0,0 +1,88 @@
+  bufferlist bl;
+  rc = client->ll_preadv_pwritev(fh, iov_out, 2, 0, true, writefinish.get(), nullptr);
+  ASSERT_EQ(0, rc);
+  rc = writefinish.get()->wait();
```
Ah, OK, yes that works. Will update with that.

ffilz commented on src/test/client/async.cc:
```diff
@@ -0,0 +1,88 @@
+
+#include "test/client/TestClient.h"
+
+#if 1
```
OK

ffilz commented on src/test/client/async.cc:
```diff
@@ -0,0 +1,88 @@
+ * License version 2.1, as published by the Free Software
+ * Foundation.  See file COPYING.
+ *
+ */
```
Needs to be changed to 2022

ffilz replied:
fixed

lxbsz commented on src/client/Client.cc:
```diff
@@ -9823,18 +9823,180 @@ int Client::preadv(int fd, const struct iovec *iov, int iovcnt, loff_t offset)
   utime_t start = ceph_clock_now(); 
+  CRF_onuninline *crf_onuninline = NULL;
+  CRF_iofinish *crf_iofinish = NULL;
+  C_SaferCond *cond_onuninline = NULL;
```
`s/NULL/nullptr/`

would be idiomatic c++

ffilz replied:
fixed

ffilz commented on src/client/Client.cc:
```diff
@@ -10055,16 +10067,63 @@ int64_t Client::_read(Fh *f, int64_t offset, uint64_t size, bufferlist *bl)
     if (rc < 0)
       goto done;
+  } else if (onfinish) {
+    // handle _sync_read asynchronously...
```
I added comments to hopefully make this make more sense... There is a method, Client::_read_sync which is a code path the async (non-blocking) path still needs to follow, but it can't actually call _read_sync() to do it, thus the C_Read_Sync_Async class to manage the asynchronous non-blocking execution of that code path...

Maybe there's a better way to name it, I just wanted to help tie it into the blocking code path it emulates...

ajarr replied:
thanks for the comments. I'm able to understand this better.

ffilz commented on src/client/Client.h:
```diff
@@ -632,6 +632,9 @@ class Client : public Dispatcher, public md_config_obs_t {
   int ll_write(Fh *fh, loff_t off, loff_t len, const char *data);
   int64_t ll_readv(struct Fh *fh, const struct iovec *iov, int iovcnt, int64_t off);
   int64_t ll_writev(struct Fh *fh, const struct iovec *iov, int iovcnt, int64_t off);
+  int ll_preadv_pwritev(struct Fh *fh, const struct iovec *iov, int iovcnt,
```
Fixed, though I note there are other callers of _preadv_pwritev_locked which turn the result into an int (that probably are why I coded as int...), but I'll leave that for someone else to change...

ffilz commented on src/include/cephfs/libcephfs.h:
```diff
@@ -1876,6 +1888,8 @@ int64_t ceph_ll_readv(struct ceph_mount_info *cmount, struct Fh *fh,
 int64_t ceph_ll_writev(struct ceph_mount_info *cmount, struct Fh *fh,
 		       const struct iovec *iov, int iovcnt, int64_t off);
+int64_t ceph_ll_async_rw(struct ceph_mount_info *cmount,
+			 struct ceph_ll_io_info *io_info);
```
I went with ceph_ll_async_readv_writev. If someone else needs non-blocking buffer based API, they can add it...

ffilz commented on src/client/Client.h:
```diff
@@ -1222,6 +1222,213 @@ class Client : public Dispatcher, public md_config_obs_t {
+      iofinished_r = 0;
+      onuninlinefinished_r = 0;
+      iofinished = false;
+      onuninlinefinished = onuninline == nullptr;
```
We don't always have to uninline a file. If so, there will be no onuninline Context, so we need to mark it already finished.

lxbsz replied:
@ajarr means the type of `onuninlinefinished` is `bool`, you should assign it to `false` instead of `nullptr`.

ffilz replied:
Right, onuninlinefinished is bool, It isn't being assigned onuninline

It's being assigned the result of (onuninline == nullptr). The == operator results in a bool.

ajarr replied:
Ah OK!  It looked like onunlinefinished = onunline = false. Thanks!

Yes, a comment mentioning "We don't always have to uninline a file. If so, there will be no onuninline Context, so we need to mark it already finished." should also help.

ffilz commented on src/client/Client.h:
```diff
@@ -1222,6 +1222,213 @@ class Client : public Dispatcher, public md_config_obs_t {
+      iofinished_r = 0;
+      onuninlinefinished_r = 0;
+      iofinished = false;
+      onuninlinefinished = onuninline == nullptr;
```
We don't always have to uninline a file. If so, there will be no onuninline Context, so we need to mark it already finished.

I should comment to that effect though...

ajarr replied:
Got it. A comment will help.

ajarr reviewed (commented):
I went through Client::_read() in detail. I've a few more questions/comments.

I am finding Client::C_Read_Sync_Async::finish() tricky. I'm still going through that couple of more times.

I also need to go through fsync related changes.

ffilz commented on src/client/Client.cc:
```diff
@@ -9823,18 +9823,184 @@ int Client::preadv(int fd, const struct iovec *iov, int iovcnt, loff_t offset)
+{
+  bool fini;
+
+  clnt->client_lock.lock();
```
Hmm, uninline_data can complete immediately, clearly with the client_lock held.

If it actually runs, it's not clear to me if the client_lock is held or not (its all done in objector), I'm thinking not.

The original onuninline using a C_SaferCond doesn't need the client lock to signal or wait for the condition, so doesn't care if client_lock is held or not...

Considering we really shouldn't have to uninline data, maybe I should just make uninline synchronous, and always use a C_SaferCond and wait for it to be done before returning, the _read (for example) calls finish_onunline with the client_lock held.

ffilz replied:
Cool, I have done that then.

lxbsz commented on src/client/Client.cc:
```diff
@@ -9873,7 +10039,13 @@ int64_t Client::_read(Fh *f, int64_t offset, uint64_t size, bufferlist *bl)
+      } else {
+        cond_onuninline = new C_SaferCond("Client::_read_uninline_data flock"); 
+        onuninline.reset(cond_onuninline);
+      }
```
It's buggy here. Not from your patches.

We shouldn't uninline the data when reading, because it won't always be granted the `Fw` caps.

Fixed in https://github.com/ceph/ceph/pull/47090.

ffilz replied:
I've rebased onto this patch so I don't need to deal with unlinline for read.

ffilz commented on src/client/Client.cc:
```diff
@@ -9823,18 +9823,184 @@ int Client::preadv(int fd, const struct iovec *iov, int iovcnt, loff_t offset)
+  // Caller holds client_lock so we don't need to take it.
+
+  if (r >= 0) {
+    if (is_read_async) {
```
Ah, for that one, I had a question if I needed to do anything to deal with a short read

Handling would be different depending on whether read is sync or async (a separate sync/async from what I'm doing...)

ffilz commented:
There is some confusion about use of sync and async in existing code. I'm thinking maybe I need to change the wording of my efforts from async to nonblocking.

ffilz commented on src/client/Client.cc:
```diff
@@ -9823,18 +9823,184 @@ int Client::preadv(int fd, const struct iovec *iov, int iovcnt, loff_t offset)
+ */
+void Client::C_Read_Sync_Async::finish(int r)
+{
+  clnt->client_lock.lock();
```
finish will be called via read_trunc, which looks like it is scheduled so runs on another thread. I haven't followed that code path all the way to the gory end. Maybe it does take the client_lock somehwere? But it all goes down into filer and objecter so seems unlikely any of that would acquire the client lock, and I don't see any inline completion that could call finish on the submitting thread that WOULD still hold the client_lock.

I welcome more analysis on this is one place I got in over my head on following and understanding the code.

ffilz commented on src/client/Client.cc:
```diff
@@ -9823,18 +9823,184 @@ int Client::preadv(int fd, const struct iovec *iov, int iovcnt, loff_t offset)
+    }
+
+    ceph_assert(r >= 0);
+    if (movepos) {
```
Yes, that will be fixed in the next update.

ffilz commented on src/client/Client.cc:
```diff
@@ -10381,36 +10570,46 @@ int Client::_read_async(Fh *f, uint64_t off, uint64_t len, bufferlist *bl)
+    }
+    return 0;
+  }
+
```
No, it's a managed pointer (std::unique_ptr). It will go out of context and be cleaned up on any exit from Client::_read_async(). We release it HERE because it needs to have a life beyond the call to _read_async().

ffilz commented on src/client/Client.cc:
```diff
@@ -10224,6 +10219,51 @@ int Client::_preadv_pwritev(int fd, const struct iovec *iov, unsigned iovcnt, in
+  // time
+  lat = ceph_clock_now();
+  lat -= start;
+  logger->tinc(l_c_wrlat, lat);
```
Thanks for that, I put that change in.

ffilz commented on src/client/Client.cc:
```diff
@@ -9823,18 +9823,184 @@ int Client::preadv(int fd, const struct iovec *iov, int iovcnt, loff_t offset)
+    
+    lat = ceph_clock_now();
+    lat -= start;
+    clnt->logger->tinc(l_c_read, lat);
```
Thanks for that, I put that change in.

ffilz commented:
OK, I believe this update includes all the required fixes.

I'm still not sure if I need any short read handling in C_Read_Finisher::finish_io.

The client_lock issue in C_Read_Sync_Async::finish still needs more analysis.

ffilz commented:
I posted a new version changing the title and names from async to nonblocking to reduce confusion.

Here is a pastebin of the diff:

https://pastebin.com/MU7ZtLaL

Here is the nonblocking-io branch:

https://github.com/ffilz/ceph/commits/nonblocking-io

If folks think this looks good, I will push the changes to this pull request (and update the pull request title).

I think this makes things easier to understand and gets rid of Sync_Async in function names...

lxbsz commented:
@ffilz I suggest just wait for @mchangir uninline_data PR#44359, , which will add a scrub command to uninline the inline_data, and a following PR which will remove the inline data code from `Client.cc`.

Or in this PR you can just skip the uninline related code by adding some comments or just return not supported if it has inline data ?

ffilz commented:
Dear reviewers,

I need to figure out one more thing, and maybe I need to drop the client_lock to make the callback to Ganesha. I also possibly have to close the filehandle used. For ganesha to call ceph_ll_close, the client_lock can't be held since ll_close will take it.

It was reasonable to add non-blocking fsync on write completion, but file close at end of I/O may be too much to ask... (plus it's a bit tricky in the Ganesha code since the callback isn't actually managing the lifetime of the filehandle).

Can we safely drop the client_lock in the callback?

ffilz commented:
Please see new pull request https://github.com/ceph/ceph/pull/48038
